twister: get information about memory footprint from build.log

katgiadla opened the pull request:
According #2465, memory foootprint for test configuration is read from _build.log_.

Using build.log report, it is possible to get more accurate information about usage of RAM/ROM memory.

These changes calculate memory footprint from build.log add this information to JSON report.

katgiadla committed:
twister: get information about memory footprint from build.log

This commit calculate memory footprint from build.log and
proposes an alternative approach to #2465.

Signed-off-by: Katarzyna Giądła <katarzyna.giadla@nordicsemi.no>

gopiotr reviewed (changes requested):
First of all: thank you for your contribution :)
In general I agree that those changes are necessary to make memory calculation more accurate.
I think that all those calculations should be moved into `SizeCalculator` class in `size_calc.py` file
https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/size_calc.py#L12.
`TestInstance` class is not good place to do this due to the fact that we already have dedicated class for this purpose and thanks to this we can maintain better separation of functionalities. It is not good idea to maintain two "version" of size calculation in Twister - old one should be replaced by new one.

PerMac reviewed (changes requested):
Please don't introduce new libraries if they are not crucial for the execution, i.e. pandas in this case. I believe all the calculations from the PR can be done without such advanced library and data types it introduces.  IMO it will also make the code more understandable for people not familiar with this lib. 

I've yet to do a full review.

katgiadla commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -268,8 +307,144 @@ def calculate_sizes(self):
+        try:
+            with open(file=path, mode='r') as file:
+                content_file = file.readlines()
+        except(Exception) as e:
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -268,8 +307,144 @@ def calculate_sizes(self):
+                break
+        #If file does not contain information about memory footprint, warning is raised.
+        if result == 0:
+            warnings.warn(message=MSG_NOT_FOUND_FOOTPRINT)
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -268,8 +307,144 @@ def calculate_sizes(self):
+
+        @return typing.List[str]: Lines with information about memory footprint.
+        """    
+        return file_content[(len(file_content) - (start_idx+1)):(len(file_content) - (start_idx+1) + AMOUNT_USEFUL_LINES)]
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -268,8 +307,144 @@ def calculate_sizes(self):
+            log_path = os.path.join(self.build_dir, path)
+            try:
+                file_content = self.get_log_file_content(path=log_path)
+            except (Exception) as e:
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -268,8 +307,144 @@ def calculate_sizes(self):
+            except (Exception) as e:
+                print(e.with_traceback)
+            data_start_line_idx = self.search_idx_of_last_pattern_occurrence(file_content=file_content, pattern=PATTERN_SEARCHED_LINE)
+            data_from_file = self.create_data_frame(lines=self.get_lines_with_footprint(start_idx=data_start_line_idx, file_content=file_content))
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -10,12 +10,51 @@
+# Default exponent to binary convert
+EXPONENT_DEFAULT = 10
+# Basis of the binary power
+BINARY_MULTIPIER_BYTES = 2
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -10,12 +10,51 @@
+MSG_NOT_FOUND_FOOTPRINT = "Information about memory footprint for this test configuration is not found."
+
+# Order of the magnitude by units
+UNITS_MULTIPLIERS = {'B': 0, \
```
Done

PerMac commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +260,154 @@ def _calculate_sizes(self):
+
+        return data_lines
+
+    def _convert_value_to_unit(self, value: str, unit_out: str) -> str:
```
_binary_prefix_converter?

katgiadla replied:
Done

PerMac commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +277,155 @@ def _calculate_sizes(self):
+            content_file = []
+        return content_file
+
+    def _search_idx_of_last_pattern_occurrence(self, file_content: typing.List[str], pattern: str) -> int:
+        """Search number of line with information about memory footprint.
+
+        @param file_content (str) Content of build.log.
+        @param pattern (str) String with start phrase of memory footprint section.
+        @return: Number of offset with information about the memory footprint (int)
+        """
+        result = 0
+        if not pattern:
+            pattern = self.PATTERN_SEARCHED_LINE
+
+        # Check file from tail
+        for idx, line in enumerate(reversed(file_content)):
+            # Condition is fullfied, if pattern is matched with start of line.
+            if re.match(pattern=pattern, string=line):
+                result = idx
+                break
+        # If file does not contain information about memory footprint, warning is raised.
+        if result == 0:
+            logger.error(msg=self.MSG_NOT_FOUND_FOOTPRINT)
+        return result
```
Please unify wording in comments and variables names: offset, number of a line, index etc.

katgiadla replied:
Done

katgiadla commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -831,31 +832,39 @@ def run(self):
         if self.options.enable_size_report and not self.options.cmake_only:
-            self.calc_one_elf_size(instance)
+            # choice source of the getting the memory footprint
+            # check if this information should be read from build.log
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -91,49 +97,42 @@ class SizeCalculator:
+    ROW_ROM_IDX = 1
+    COLUMN_USED_SIZE_IDX = 1
+    COLUMN_PERCENTAGE_USAGE_IDX = 3
+    MSG_NOT_USEFUL_INFO = "Missing information about the memory footprint. Please check build.log."
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -160,18 +159,72 @@ def get_rom_size(self):
+        else:
+            logger.error("Incorrect file to analyse of the footprints.")
+
+    # Separate ELF functionalities from __init__()
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +277,155 @@ def _calculate_sizes(self):
+            with open(file=self.filename, mode='r') as file:
+                content_file = file.readlines()
+        else:
+            logger.error("Incorrect build.log file to analyse of the footprints.")
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -256,21 +255,39 @@ def create_overlay(self, platform, enable_asan=False, enable_ubsan=False, enable
 
         @return A SizeCalculator object
         """
+        # choice source of the getting of the memory footprint
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -91,49 +97,39 @@ class SizeCalculator:
+    EXPONENT_DEFAULT = 10
+    # Basis of the binary power
+    BINARY_MULTIPLIER_BYTES = 2
+    UNITS_MULTIPLIERS = {'B': 0, 'KB': 1, 'MB': 2, 'GB': 3}
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +277,155 @@ def _calculate_sizes(self):
+
+        if not data_from_file:
+            pass
+        else:
```
Done

hakehuang reviewed (dismissed):
looks good to me

gopiotr commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -160,18 +147,71 @@ def get_rom_size(self):
+                    slist.append(v["name"])
         return slist
 
     def _calculate_sizes(self):
```
When Twister is used with `--from-buildlog` argument, then the verification of unrecognized sections occurrence is completely disabled and following fail will never occur: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/reports.py#L417-L421. It should be at least mentioned somewhere in `--from-buildlog` argument help description.

I wondering if the best solution for this problem is execute this elf file scanning every time (and thanks to this sections will be still monitored). And when user would like to use `--from-buildlog`, then `rom` and `ram` sizes would be calculated once again based this time on `build.log`. Thanks to this approach `sections` list will not be empty and could be used to properly verification of `unrecognized_sections`.

katgiadla replied:
Done

gopiotr commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -160,18 +147,71 @@ def get_rom_size(self):
 
     def _calculate_sizes(self):
+        if self.filename.endswith("build.log"):
+            self._get_footprint_from_buildlog(pattern_information_line=self.PATTERN_SEARCHED_LINE)
```
`pattern_information_line` is passed multiple times from one method to another. Please consider if this will be simpler if information about this pattern could be stored just as class attribute and initialize in `__init__`:
```python
def __init__(self, filename, extra_sections)
    ...
    self.pattern_information_line = self.PATTERN_SEARCHED_LINE
    ...
```

katgiadla replied:
Done

gopiotr commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +264,169 @@ def _calculate_sizes(self):
+                break
+        # If the file does not contain information about memory footprint, the warning is raised.
+        if result == 0:
+            logger.error(msg="Information about memory footprint for this test configuration is not found.")
```
What happen next, when this patter will not be found? As I understand `_find_idx_of_last_pattern_occurrence` will return index number as 0 what is unfortunately valid number for indexing and as a result program will try to continue calculations basing on this index. Maybe by default `result` variable should have value `None` what explicitly points that index was not found. Then in `_create_data_table` method should be applied extra verification: if found index is `None`, then calculation should be stopped.

katgiadla replied:
Done

gopiotr commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +264,169 @@ def _calculate_sizes(self):
+        @return: Lines with information about memory footprint (list[str])
+        """
+        info_line_idx_start = len(file_content) - (start_idx+1)
+        info_line_idx_stop = info_line_idx_start + self.USEFUL_LINES_AMOUNT
```
Please add verification if this index does not exceeds the `file_content` size - if yes, then probably calculation should be stopped, but at least it should be prevented to trying get data out of list in next line, for example:
```python
info_line_idx_stop = info_line_idx_stop if info_line_idx_stop < len(file_content) else len(file_content) - 1
```

katgiadla replied:
Done

gopiotr reviewed (changes requested):
Thank you for moving footprint calc basing on build log into `SizeCalculator` class :)

PerMac reviewed (changes requested):
commit massage states: This commit calculate memory footprint from build.log and
fixes https://github.com/zephyrproject-rtos/zephyr/issues/2465.
Please remove the fixes part. It doesn't fix the original issue. It proposes an alternative approach

katgiadla commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -601,6 +603,12 @@ def parse_arguments(args):
              "All other command line arguments ignored.")
 
+    parser.add_argument(
+        "--from-buildlog",
```
This argument cannot use without the flag `--show-footprint`. Therefore, adding the word `footprint` to this argument could be redundant.

katgiadla commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -831,31 +832,36 @@ def run(self):
 
     @staticmethod
-    def calc_one_elf_size(instance):
+    def calc_size(instance: TestInstance, should_from_buildlog: bool):
```
DOne

nashif commented:
getting this on `./scripts/twister -p qemu_x86 -T samples/hello_world/ --show-footprint --from-buildlog`

```
WARNING:root:Missing information about the memory footprint. Please check build.log.
WARNING - Missing information about memory footprints.
WARNING:twister:Missing information about memory footprints.
DEBUG:twister:run test: qemu_x86/samples/hello_world/sample.basic.helloworld
DEBUG:twister:Spawning QEMUHandler Thread for qemu_x86/samples/hello_world/sample.basic.helloworld
DEBUG:twister:Running qemu_x86/samples/hello_world/sample.basic.helloworld (qemu)
DEBUG:twister:Spawning QEMUHandler Thread for qemu_x86/samples/hello_world/sample.basic.helloworld
DEBUG:twister:QEMU (2935191): SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
DEBUG:twister:QEMU (2935191): Booting from ROM..
DEBUG:twister:QEMU (2935191): *** Booting Zephyr OS build zephyr-v3.2.0-1450-ga488e429f48f ***
DEBUG:twister:QEMU (2935191): Hello World! qemu_x86
DEBUG:twister:QEMU (2935191) complete (passed) after 2.120734214782715 seconds
DEBUG:twister:No timeout, return code from QEMU (None): 0
DEBUG:twister:return code from QEMU (None): 0
DEBUG:twister:run status: qemu_x86/samples/hello_world/sample.basic.helloworld passed

gopiotr commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +275,174 @@ def _calculate_sizes(self):
+
+            info_line_idx_stop = info_line_idx_start + self.USEFUL_LINES_AMOUNT
+            if info_line_idx_stop > len(file_content):
+                info_line_idx_stop = len(file_content) - 1
```
Should be without -1:
```python
info_line_idx_stop = len(file_content)
```
Sorry, for this - it was my suggestion in previous comment and there was a bug. In Python it is possible to use in index range number like `len(mylist)`, for example:
```python
mylist = [1, 2, 3, 4]
print(mylist[2:len(mylist)])
# what is the same as:
print(mylist[2:4])
# or simpler:
print(mylist[2:])
```

katgiadla replied:
Done

gopiotr commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +275,174 @@ def _calculate_sizes(self):
+        @return: Lines with information about memory footprint (list[str])
+        """
+        if len(file_content) == 0:
+            logging.warning(msg="Build.log file is empty.")
```
Please change it from `logging` into `logger`

katgiadla replied:
Done

gopiotr commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +275,174 @@ def _calculate_sizes(self):
+        """
+        if len(data_lines) != self.USEFUL_LINES_AMOUNT:
+            data_lines = [[]]
+            logging.warning(msg="Missing information about the memory footprint. Please check build.log.")
```
Please change it from `logging` into `logger`

PerMac commented:
@nashif seems that it was a "typo" issue logging->logger. However, what do you think should be the behavior for qemu/native_posix builds? They by default don't contain this information (qemu has only ram and native none). Should the warning be turned off for such platforms or left?

nashif commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +259,169 @@ def _calculate_sizes(self):
+                    break
+        # If the file does not contain information about memory footprint, the warning is raised.
+        if result == -1:
+            logger.warning(msg="Information about memory footprint for this test configuration is not found.")
```
this will be a lot of warnings, all qemu tragets will have this issue and the warning does not say anything about where the issue is anyways, I would turn this into debug and add more information about where the issue is.

```
WARNING - Missing information about the memory footprint. Please check build.log.
WARNING - Missing information about memory footprints.
WARNING - Missing information about the memory footprint. Please check build.log.
WARNING - Missing information about memory footprints.
WARNING - Missing information about the memory footprint. Please check build.log.
WARNING - Missing information about memory footprints.
WARNING - Missing information about the memory footprint. Please check build.log.
WARNING - Missing information about memory footprints.
WARNING - Missing information about the memory footprint. Please check build.log.
WARNING - Missing information about memory footprints.
WARNING - Missing information about the memory footprint. Please check build.log.
WARNING - Missing information about memory footprints.
WARNING - Missing information about the memory footprint. Please check build.log.
WARNING - Missing information about memory footprints.
````

evgeniy-paltsev replied:
If it is produced for QEMU I'd say it shouldn't be even with debug verbosity.

katgiadla replied:
Comments change to clarify the source of the problem.

katgiadla replied:
There are several versions of QEMU-type platforms that differ in architecture. Depending on which platform you choose, you either do not get information about the FLASH memory used, or you get information about the SRAM memory (instead of RAM). This information may not be valuable. 
For this reason, memory footprints are not taken for QEMUs and warnings are not displayed for them.

nashif reviewed (changes requested):
I am surprised how many helper functions you add just to read some data from a file, this just makes reading the code which should be straightforward very difficult (jumping between helper functions that are being used only once). 
Oh well, as long as it works...

katgiadla commented on scripts/pylib/twister/twisterlib/reports.py:
```diff
@@ -248,29 +248,32 @@ def json_report(self, filename, version="NA"):
             handler_time = instance.metrics.get('handler_time', 0)
             ram_size = instance.metrics.get ("ram_size", 0)
             rom_size  = instance.metrics.get("rom_size",0)
+            percentage_usage_ram = instance.metrics.get("percentage_usage_ram", 0)
```
I think getting information about the percentage of memory usage is necessary. The build.log file contains information about the rate RAM/FLASH usage calculated based on the real available memory. Therefore, it is no need to calculate memory usage additionally. Given this fact, I consider the addition of information on available memory to the report redundant

PerMac replied:
@nashif I am not sure if we are on the same page. This is an example of information available in the build.log. 
``` 
[144/144] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       61588 B         1 MB      5.87%
             RAM:          9 KB       256 KB      3.52%
        IDT_LIST:          0 GB         2 KB      0.00%
  ```
 By available memory, you mean the `Region size` or memory left, i.e. `region size - used size`? 
I think memory left can be useful, since it can help with understanding how close to the edge we are, in the terms of absolute values (kB left). Not sure why Region Size would be useful to store in the report for every image, can you elaborate if this is what you prefer?

>  percentage can then be calculated from used and available if needed.

We want to have the percentage in the report, since it was already calculated and prefer not to have to recalculate it again, after the report is ready. Such json report can then be directly feed to the data base (elastic search, grafana, etc). We found monitoring of the percentage useful and are planning to have an internal triggers which will warn us when % raise above a given value for our samples and applications.

katgiadla replied:
@nashif thanks for the clarification. Code has been changed.

katgiadla commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -256,21 +255,36 @@ def create_overlay(self, platform, enable_asan=False, enable_ubsan=False, enable
         """
+        elf_filepath = self.get_elf_file()
+        buildlog_filepath = self.get_buildlog_file() if from_buildlog else ''
+        return SizeCalculator(elf_filename=elf_filepath, extra_sections=self.testsuite.extra_sections, buildlog_filepath=buildlog_filepath)
```
Analyzing the ELF file is still necessary because of the search for potential errors (see https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/reports.py#L417-L421). This is the reason why we cannot exclude analyzing the ELF file. Consequently, in the scenario of using the _build.log_ file, it is necessary to pass the paths to both files.
Given the logic and architecture of Twister, the most optimal solution is to pass both paths still in the `TestInstance` class. In the `SizeCalculator` class, the existence of a path to build.log checks. If _build.log_ has been passed and exists, then the procedure of reading from _build.log_ begins.

katgiadla commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +259,169 @@ def _calculate_sizes(self):
+
+        @return Content of the build.log file (list[str])
+        """
+        if os.path.exists(path=self.buildlog_filename) and self.buildlog_filename.endswith("build.log"):
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/testinstance.py:
```diff
@@ -256,21 +255,36 @@ def create_overlay(self, platform, enable_asan=False, enable_ubsan=False, enable
         """
+        elf_filepath = self.get_elf_file()
+        buildlog_filepath = self.get_buildlog_file() if from_buildlog else ''
+        return SizeCalculator(elf_filename=elf_filepath, extra_sections=self.testsuite.extra_sections, buildlog_filepath=buildlog_filepath)
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -224,3 +259,169 @@ def _calculate_sizes(self):
+    def _get_footprint_from_buildlog(self) -> None:
+        """Get memory footprint from build.log.
+
+        @param pattern_information_line (str) Pattern with first line of the information
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -91,51 +97,31 @@ class SizeCalculator:
     ]
 
-    def __init__(self, filename, extra_sections):
+    # Variables below are stored for calculate size using build.log
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -91,51 +97,31 @@ class SizeCalculator:
+    # Pattern to first line with information about memory footprint
+    PATTERN_SEARCHED_LINE = "Memory region"
+
+    def __init__(self, elf_filename: str, extra_sections: typing.List[str], buildlog_filepath: str = ''):
```
Done

katgiadla commented on scripts/pylib/twister/twisterlib/size_calc.py:
```diff
@@ -205,17 +231,26 @@ def _calculate_sizes(self):
             if name in SizeCalculator.alloc_sections:
-                self.ram_size += size
-                stype = "alloc"
+                if self.buildlog_filename:
```
Done

katgiadla commented:
@nashif  I've addressed all your comments:
- used and available memory are taken into a report, percentage is 
- not warnings about missing footprints are turned off for native posix and qemu. These platforms won't report memory info at all
- changed name of the flag

Sizce_calc still uses both elf and `build.log`, since elf is used for detection of unrecognized regions.

nashif commented:
Do we really need decimals?
```
            "rom_size":62784.0,
            "retries":0,
            "available_ram":262144.0,
            "available_rom":262144.0,
```

katgiadla commented:
@nashif thanks for your points. Your suggestions of the changes are applied.
